README.md

# Deployment steps

* Clone the repo.

* Create `.env` file with the following variables and required values:

  ```env
  RABBITMQ_HOSTNAME=192.254.254.254
  RABBITMQ_USERNAME=guest
  RABBITMQ_PASSWORD=guest

  HOST_XFS_VOLUME=/some/path
  CONTAINER_NAME=Some-Docker-Container-Name
  ```

  * `RABBITMQ_HOSTNAME`, `RABBITMQ_USERNAME`, and `RABBITMQ_PASSWORD` provide details needed to access the rabbitmq.
  * `HOST_XFS_VOLUME` is a path on the host that will run the automated code. This needs to refer to the location where overseer containers will store their work. Overseer uses this to setup a volume in each container it creates. When multiple overseer instances are run this path needs to be unique for each instance. Overseer will create `sandbox` and `output` folders in this location.
  * `CONTAINER_NAME` is optional (defaults to `overseer-container`) and provides the name of the container that overseer creates. If there are multiple overseer instances then they must have unique container names.

* Easiest option is to setup rabbitmq alongside overseer in the same docker network, and have it setup to run with the management interface.
  
* Alternatively:
  * Configure your machine to add an alias IP address to en0 (or whatever network device you are using) as the value specified to the key `RABBITMQ_HOSTNAME` in the .env file before. On MacOS, you can use: `sudo ifconfig en0 alias 192.254.254.254 255.255.255.0`
  * Run RabbitMQ server:

    `docker run -d --hostname my-rabbit --name some-rabbit -p 5672:5672 -p 15672:15672 rabbitmq:3`

  * Enable RabbitMQ management plugin:

    `docker exec -t some-rabbit rabbitmq-plugins enable rabbitmq_management`

* Run overseer using:
    `bundle exec ruby ./app.rb`

# Basic operation

* When overseer starts is listens to the **q.tasks** queue on the **ontrack** exchange of the rabbitmq and awaits any messages marked as **task.submission**. When a submission is received the message triggers action in overseer.
* Params received include:
  * `output_path`: where to store automated assessment results
  * `submission`: the path to the submission zip file or folder
  * `assessment`: the path to the submission assessment resources
  * `timestamp`: the time of the submission
  * `task_id`: the task associated with the submission
  * `overseer_assessment_id`: the overseer assessment id
  * `docker_image_name_tag`: the tag of the docker image to run the scripts within
* Overseer starts by unpacking the submission along with the assessment resources to the working directory. Files in the assessment resources will override those in the submission if needed.
* Once the files are in place Overseer starts a container with the indicated `CONTAINER_NAME` using the image details from the assessment resources and runs two scripts: `build.sh` is run and if the result is 0 it runs `run.sh`.
  * Anything written to standard output and standard error from these scripts will be collected in the output message for the task.
  * Each of these scripts is passed a path to a yaml file where comment and status changes are stored.
  * `build.sh` should write messages to the `build_message` key in the yaml file.
  * `run.sh` should write messages to the `run_message` key in the yaml file.
  * Both build and run scripts can write a new status to `new_status`.
  * The exit status of the build and run steps is appended to the output text collected from standard out.
* When both scripts have been run, a message is returned to Doubtfire/OnTrack using the `overseer.result` routing key to the `q.overseer` of the `ontrack` exchange. This contains the task id, timestamp, and overseer assessment id. The output text and yaml files are also available to OnTrack.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [6.0.2](https://github.com/macite/doubtfire-deploy/compare/v6.0.1...v6.0.2) (2023-04-11)


### Bug Fixes

* add registry to image names ([92b6745](https://github.com/macite/doubtfire-deploy/commit/92b67453a1a7654a18d56a11c5caff5d434223b5))
* enable oversser on arm machines ([f1058c1](https://github.com/macite/doubtfire-deploy/commit/f1058c17335b8dc57cf0f2dab497c9565827091f))

### [6.0.1](https://github.com/macite/doubtfire-deploy/compare/v6.0.0...v6.0.1) (2022-03-22)

## [6.0.0](https://github.com/macite/doubtfire-deploy/compare/v5.0.8...v6.0.0) (2022-02-02)


### Features

* update overseer to ruby 3.1 ([f1bf9f7](https://github.com/macite/doubtfire-deploy/commit/f1bf9f71ed78a7fc6efb0f5f6e858d2f6c336829))

### [5.0.6](https://github.com/macite/doubtfire-deploy/compare/v5.0.5...v5.0.6) (2021-12-06)

### [5.0.5](https://github.com/macite/doubtfire-deploy/compare/v5.0.4...v5.0.5) (2021-11-19)


### Bug Fixes

* update bunny pub sub to fix launch and logging issues ([b9d967a](https://github.com/macite/doubtfire-deploy/commit/b9d967adbdf6ae1a36dd218c77c64d03c1fea09a))

### [5.0.4](https://github.com/macite/doubtfire-deploy/compare/v5.0.3...v5.0.4) (2021-11-19)


### Bug Fixes

* update bunny pub sub to fix subscriber launch ([57aeca6](https://github.com/macite/doubtfire-deploy/commit/57aeca6a92f94a539d0449de139dd8a841e9ab38))

### [5.0.3](https://github.com/macite/doubtfire-deploy/compare/v5.0.2...v5.0.3) (2021-11-19)


### Bug Fixes

* remove some limits to ensure test containers run ([9daab39](https://github.com/macite/doubtfire-deploy/commit/9daab3964fd75e218cd7e5f4ca2a44ccf994b1f6))

### [5.0.2](https://github.com/macite/doubtfire-deploy/compare/v5.0.1...v5.0.2) (2021-11-18)


### Features

* add proxy login to overseer ([9435c79](https://github.com/macite/doubtfire-deploy/commit/9435c79b1690153cecbb046542d7ef4d0990f452))

### [5.0.1](https://github.com/macite/doubtfire-deploy/compare/v5.0.0-2...v5.0.1) (2021-10-27)

## [5.0.0](https://github.com/macite/doubtfire-deploy/compare/v5.0.0-2...v5.0.0) (2021-10-13)


app.rb
# frozen_string_literal: true

require 'dotenv/load'
require 'bunny-pub-sub/subscriber'
require 'bunny-pub-sub/publisher'
require_relative 'overseer_receive_action.rb'

docker_config = {
  DOCKER_PROXY_URL: ENV['DOCKER_PROXY_URL'],
  DOCKER_TOKEN: ENV['DOCKER_TOKEN'],
  DOCKER_USER: ENV['DOCKER_USER']
}
subscriber_config = {
  RABBITMQ_HOSTNAME: ENV['RABBITMQ_HOSTNAME'],
  RABBITMQ_USERNAME: ENV['RABBITMQ_USERNAME'],
  RABBITMQ_PASSWORD: ENV['RABBITMQ_PASSWORD'],
  EXCHANGE_NAME: 'ontrack',
  DURABLE_QUEUE_NAME: 'q.tasks',
  BINDING_KEYS: 'task.submission',
  DEFAULT_BINDING_KEY: 'task.submission'
}

assessment_results_publisher_config = {
  RABBITMQ_HOSTNAME: ENV['RABBITMQ_HOSTNAME'],
  RABBITMQ_USERNAME: ENV['RABBITMQ_USERNAME'],
  RABBITMQ_PASSWORD: ENV['RABBITMQ_PASSWORD'],
  EXCHANGE_NAME: 'ontrack',
  DURABLE_QUEUE_NAME: 'q.overseer',
  # Publisher specific key
  # Note: `*.result` works too, but it makes no sense using that.
  ROUTING_KEY: 'overseer.result'
}

if docker_config[:DOCKER_TOKEN] && docker_config[:DOCKER_PROXY_URL]
  puts "Logging into docker proxy"
  `echo \"${DOCKER_TOKEN}\" | docker login --username ${DOCKER_USER} --password-stdin ${DOCKER_PROXY_URL}`
end

assessment_results_publisher = Publisher.new assessment_results_publisher_config

# Register subscriber for task submissions, runs overseer receive, and publishes results to assessment_results_publisher
register_subscriber(subscriber_config,
                    method(:receive),
                    assessment_results_publisher)


receive_rabbit_mq.rb
require "bunny"
require "json"
require "zip"
require "pathname"

class ClientException < RuntimeError
  attr_reader :status

  def initialize(status = 400, message)
    puts "Client error: #{message.to_s}, #{status}"
    @status = status
    super(message)
  end
end

class ServerException < RuntimeError
  attr_reader :status

  def initialize(status = 500, message)
    puts "Server error: #{message.to_s}, #{status}"
    @status = status
    super(message)
  end
end

def client_error!(message, status, _headers = {}, _backtrace = [])
  raise ClientException.new status, message
end

def server_error!(message, status, _headers = {}, _backtrace = [])
  raise ServerException.new status, message
end

class Receiver
  def default_return(value)
    { message: value ||= "" }
  end

  def valid_zip?(file)
    zip = Zip::File.open(file)
    true
  rescue StandardError
    false
  ensure
    zip.close if zip
  end

  # Flat extract a zip file, no sub-directories.
  def extract_zip(input_zip_file_path, output_loc)
    Zip::File.open(input_zip_file_path) do |zip_file|
      # Handle entries one by one
      zip_file.each do |entry|
        # Extract to file/directory/symlink
        puts "Extracting #{entry.name} #{entry.ftype}"
        pn = Pathname.new entry.name
        entry.extract "#{output_loc}/#{pn.basename}" unless entry.ftype.to_s == "directory"
      end
      # # Find specific entry
      # entry = zip_file.glob('*.csv').first
      # puts entry.get_input_stream.read
    end
  end

  def get_project_path(project_id)
    "student_projects/project_#{project_id}"
  end

  ##################################################################
  ##################################################################

  # Step 1 -- done
  def extract_submission(zip_file, path)
    extract_zip zip_file, path
  end

  # Step 2 -- done
  def extract_assessment(zip_file, path)
    extract_zip zip_file, path
  end

  # Step 3
  def run_assessment_script(path)
    rpath = "#{path}/run.sh"
    server_error!({ error: "File #{rpath} doesn't exist" }, 500) unless File.exists? rpath
    result = {}
    `chmod +x #{rpath}`
    # `cd #{path}`
    Dir.chdir path do
      result = { run_result: `./run.sh` }
    end
    result
  end

  # Step 4
  def cleanup_after_your_own_mess(path)
    if File.exists? path
      puts "Recursively force removing: " + path
      FileUtils.rm_rf path
    end
  end

  ##################################################################
  ##################################################################

  def start
    connection = Bunny.new(hostname: ENV['RABBITMQ_HOSTNAME'] || 'localhost', username: "guest", password: "guest")
    connection.start

    channel = connection.create_channel
    exchange = channel.topic('asssessment', :durable => true)
    #channel.prefetch(1) # Use this for making rabbitMQ not give a worker more than 1 jobs if it is already working on one.

    queue = channel.queue(ENV['ROUTE_KEY'], durable: true)
    

    language_environments = ENV['LANGUAGE_ENVIRONMENTS'].split(',')
    language_environments.each do |language_environment| # language_environments can be something like "#.csharp" "#.splashkit.csharp" "#.python", etc.
      queue.bind(exchange, routing_key: language_environment)
    end
    queue.bind(exchange, routing_key: ENV['DEFAULT_LANGUAGE_ENVIRONMENT']) unless ENV['DEFAULT_LANGUAGE_ENVIRONMENT'].nil?

    begin
      puts " [*] Waiting for messages. To exit press CTRL+C"
      queue.subscribe(manual_ack: true, block: true) do |delivery_info, _properties, params|
        params = JSON.parse(params)
        puts params

        client_error!({ error: "Invalid zip file: #{params["submission"]}" }, 400) unless valid_zip? params["submission"]
        client_error!({ error: "Invalid zip file: #{params["assessment"]}" }, 400) unless valid_zip? params["assessment"]
        output_loc = get_project_path params[:project_id]
        FileUtils.mkdir_p output_loc
        skip_rm = params[:skip_rm] || 0

        begin
          extract_submission params["submission"], output_loc
          extract_assessment params["assessment"], output_loc

          result = run_assessment_script output_loc
          puts result

          channel.ack(delivery_info.delivery_tag)
        rescue ClientException => e
          client_error!({ error: e.message }, e.status)
        rescue ServerException
          if skip_rm != 1
            cleanup_after_your_own_mess output_loc
          end
          server_error!({ error: "Internal server error" }, 500)
        rescue
          if skip_rm != 1
            cleanup_after_your_own_mess output_loc
          end
          server_error!({ error: "Internal server error" }, 500)
        else
          if skip_rm != 1
            cleanup_after_your_own_mess output_loc
          end

          default_return result
          channel.ack(delivery_info.delivery_tag)
        end # End inner begin

      end
    rescue Interrupt => _
      connection.close

      exit(0)
    end # End outer begin
  end
end


recieve_rabbit_mq.rb
require "bunny"
require "json"
require "zip"
require "pathname"

class ClientException < RuntimeError
  attr_reader :status

  def initialize(status = 400, message)
    puts "Client error: #{message.to_s}, #{status}"
    @status = status
    super(message)
  end
end

class ServerException < RuntimeError
  attr_reader :status

  def initialize(status = 500, message)
    puts "Server error: #{message.to_s}, #{status}"
    @status = status
    super(message)
  end
end

def client_error!(message, status, _headers = {}, _backtrace = [])
  raise ClientException.new status, message
end

def server_error!(message, status, _headers = {}, _backtrace = [])
  raise ServerException.new status, message
end

class Receiver

  ##################################################################
  ##################################################################

  def start
    puts ENV['RABBITMQ_HOSTNAME']
    # return
    connection = Bunny.new(hostname: ENV['RABBITMQ_HOSTNAME'] || 'localhost', username: "guest", password: "guest")
    connection.start

    channel = connection.create_channel
    exchange = channel.topic('asssessment', :durable => true)
    #channel.prefetch(1) # Use this for making rabbitMQ not give a worker more than 1 jobs if it is already working on one.

    queue = channel.queue(ENV['ROUTE_KEY'], durable: true)
    

    language_environments = ENV['LANGUAGE_ENVIRONMENTS'].split(',')
    language_environments.each do |language_environment| # language_environments can be something like "#.csharp" "#.splashkit.csharp" "#.python", etc.
      queue.bind(exchange, routing_key: language_environment)
    end
    queue.bind(exchange, routing_key: ENV['DEFAULT_LANGUAGE_ENVIRONMENT']) unless ENV['DEFAULT_LANGUAGE_ENVIRONMENT'].nil?

    begin
      puts " [*] Waiting for messages. To exit press CTRL+C"
      queue.subscribe(manual_ack: true, block: true) do |delivery_info, _properties, params|
        params = JSON.parse(params)

        puts params
        channel.ack(delivery_info.delivery_tag)
      end
    rescue Interrupt => _
      connection.close

      exit(0)
    end # Outer begin
  end
end


overseer_receive_action.rb
# frozen_string_literal: true

require 'zip'
require 'securerandom'
require 'yaml'

# CONSTANTS. Will never change.
# Start ==========================================================

module CONSTANTS
  RUN              = 'run'
  BUILD            = 'build'

  # Mounted as a volume
  DOCKER_WORKDIR   = 'home/overseer/work-dir'
  DOCKER_OUTDIR    = 'home/overseer/work-dir/output'
  DOCKER_EXECDIR   = 'home/overseer/work-dir/sandbox'
end

def docker_workdir_path
  "/#{CONSTANTS::DOCKER_WORKDIR}"
end

def docker_outdir_path
  "/#{CONSTANTS::DOCKER_OUTDIR}"
end

def docker_execdir_path
  "/#{CONSTANTS::DOCKER_EXECDIR}"
end

# End ==========================================================


# ENV CONSTANTS. Will change for different overseer instances.
# Start ==========================================================

# Used by execution container to specify host volume.
def host_xfs_volume_path
  # Root path of the directory on the HOST used as a volume for
  # temporarily storing files generated by Docker command
  # execution. Should be mounted as a XFS volume.
  ENV['HOST_XFS_VOLUME']
end

# Used by execution container to specify host volume.
def host_exec_path
  # Docker volumes needs absolute source and destination paths
  "#{host_xfs_volume_path}/sandbox"
end

# Used by execution container to specify host volume.
def host_output_path
  # Docker volumes needs absolute source and destination paths
  "#{host_xfs_volume_path}/output"
end

def container_name
  ENV['CONTAINER_NAME'] || 'overseer-container'
end

# End ==========================================================

def ack_result(results_publisher, overseer_assessment_id, task_id, timestamp, output_path)
  return if results_publisher.nil?

  msg = { overseer_assessment_id: overseer_assessment_id, task_id: task_id, timestamp: timestamp }

  results_publisher.connect_publisher
  results_publisher.publish_message msg
  results_publisher.disconnect_publisher
end

def valid_zip?(file)
  begin
    zip = Zip::File.open(file)
    return true
  rescue StandardError => e
    raise e
  ensure
    zip&.close
  end
  false
end

# Extract the zip file with all path details, or flatten based on parameters. Allows details to be overriden or not depending on parameters passed in.
def extract_zip(input_zip_file_path, output_loc, flatten = false, override = false)
  puts "Extracting:"
  puts "  zip file:".ljust(20) + input_zip_file_path
  puts "  to:".ljust(20) + output_loc
  puts "  files: "
  Zip::File.open(input_zip_file_path) do |zip_file|
    # Handle entries one by one
    zip_file.each do |entry|
      # Extract to file/directory/symlink
      unless entry.ftype.to_s == 'directory'
        if flatten
          pn = File.join(output_loc, Pathname.new(entry.name).basename)
        else
          pn = File.join(output_loc, entry.name)
        end
        override_msg = ''
        if File.exist?(pn)
          if override
            FileUtils.rm_f(pn)
            override_msg = ' OVERIDE'
          else
            puts "    - type: #{entry.ftype}".ljust(20) + " original_name: #{entry.name}".ljust(50) + " SKIPPED"
            # dont override so skip
            continue
          end
        end
        FileUtils.mkdir_p(File.dirname(pn))
        puts "    - type: #{entry.ftype}".ljust(20) + " original_name: #{entry.name}".ljust(50) + " final_name: #{pn}#{override_msg}"
        entry.extract pn
      end
    end
  end
end

def force_remove_container
  puts "Removing container forcibly: #{container_name}"
  `docker container rm -vf #{container_name}`
end

##################################################################
##################################################################

# Step 3
# Fire up a docker container to perform an execution step.
# output_path:      Destination path to write files to. This will be a OnTrack FS path.
# random_string:    Prefix for .yaml and .txt files that are to be copied.
# exec_mode:        Can either be CONSTANTS::BUILD OR CONSTANTS::RUN.
# command:          The bash command to be run via
#                   `docker run [options] <image_name_tag> /bin/bash -c "#{command}"`.
# image_name_tag:   Name and tag of the image to be run as a container.
def run_assessment_script_via_docker(output_path, random_string, exec_mode, command, image_name_tag, task_id, timestamp, overseer_assessment_id)
  client_error!({ error: "A valid Docker image_name:tag is needed" }, 400) if image_name_tag.nil? || image_name_tag.to_s.strip.empty?
  force_remove_container

  # Load registry details
  registry = ENV.fetch('DOCKER_PROXY_URL', nil)
  registry = "#{registry}/" if !registry.nil? && !registry.end_with?('/')
  registry = "" if registry.nil?

  # Add registry to image_name_tag
  image_name_tag = "#{registry}#{image_name_tag}"

  puts 'Running assessment container with the following configuration:'
  puts "  container_name: #{container_name}"
  puts "  image_name_tag: #{image_name_tag}"
  puts '  volumes:'
  puts "    - #{host_exec_path}:/#{CONSTANTS::DOCKER_EXECDIR}"
  puts "    - #{host_output_path}:/#{CONSTANTS::DOCKER_OUTDIR}"

  puts "ππππππππππππππππ Container '#{container_name}' execution for exec_mode: '#{exec_mode}' STARTED ππππππππππππππππππππ"
  puts 'π' * 120
  # TODO: Security:
  # Pass random filename... both `blah.txt` and `blah.yaml`
  # Permit write access ONLY to these files
  # Other security like no network access, capped execution time + resources, etc

  # test:
  # -m 100MB done
  # --stop-timeout 10 (seconds) (isn't for what I thought it was :))
  # --network none (fails reading from https://api.nuget.org/v3/index.json)
  # --read-only (FAILURE without correct exit code)
  # https://docs.docker.com/engine/reference/run/#security-configuration
  # https://docs.docker.com/engine/reference/run/#runtime-constraints-on-resources
  # -u="overseer" (specify default non-root user)

  # puts "docker run \
  # -m 100MB \
  # --memory-swap 100MB \
  # --restart no \
  # --cpus 1 \
  # --network none \
  # --volume #{host_exec_path}:/#{CONSTANTS::DOCKER_EXECDIR} \
  # --volume #{host_output_path}:/#{CONSTANTS::DOCKER_OUTDIR} \
  # --name #{container_name} \
  # #{image_name_tag} \
  # /bin/bash -c \"#{command}\""

  `timeout 300 docker run \
  --restart no \
  --cpus 1 \
  --network none \
  --volume #{host_exec_path}:/#{CONSTANTS::DOCKER_EXECDIR} \
  --volume #{host_output_path}:/#{CONSTANTS::DOCKER_OUTDIR} \
  --name #{container_name} \
  #{image_name_tag} \
  /bin/bash -c "#{command}"`

  puts 'π' * 120
  puts "ππππππππππππππππ Container '#{container_name}' execution for exec_mode: '#{exec_mode}' ENDED ππππππππππππππππππππππ"

  exitstatus = $?.exitstatus
  extract_result_files docker_outdir_path, output_path, random_string, exitstatus

  diff_result = `docker diff #{container_name}`
  extract_docker_diff_file output_path, diff_result, exec_mode

  puts "Docker run command execution status code: #{exitstatus}"

  # if exitstatus != 0
  #   result = {}
  #   result[:task_id] = task_id
  #   result[:overseer_assessment_id] = overseer_assessment_id
  #   result[:timestamp] = timestamp
  #   raise Subscriber::ServerException.new result, 500
  # end

  exitstatus
end

# Step 4
# Copy the results of docker execution saved at a HOST volume to OnTrack FS.
# s_path:           Source path on HOST to read files from.
# output_path:      Destination path to write files to. This will be a OnTrack FS path.
# random_string:    Prefix for .yaml and .txt files that are to be copied.
# exitstatus:       Exist status of the last `docker run` command.
def extract_result_files(s_path, output_path, random_string, exitstatus)
  client_error!({ error: "A valid output_path is needed" }, 400) if output_path.nil? || output_path.to_s.strip.empty?

  puts 'Extracting result file from the sandbox:'
  puts "  source:".ljust(20) + s_path
  puts "  destination:".ljust(20) + output_path
  puts "  file prefix:".ljust(20) + random_string

  # Get path to output files
  output_txt_file_name = "#{output_path}/output.txt"
  output_yaml_file_name = "#{output_path}/output.yaml"

  # Set files to input into the scripts
  input_txt_file_name = "#{s_path}/#{random_string}.txt"
  input_yaml_file_name = "#{s_path}/#{random_string}.yaml"

  # Process .txt file.
  if File.exist? input_txt_file_name
    File.open(input_txt_file_name, 'a') { |f|
      f.puts "exit code: #{exitstatus}"
    }

    if File.exist? output_txt_file_name
      to_append = File.read input_txt_file_name
      File.open(output_txt_file_name, 'a') { |f|
        f.puts ''
        f.puts to_append
      }
    else
      FileUtils.copy(input_txt_file_name, output_txt_file_name)
      `chmod o+w #{output_txt_file_name}`
    end

    # FileUtils.rm input_txt_file_name
  else
    puts "Results file: #{s_path}/#{random_string}.txt does not exist"
  end

  # Process .yaml file.
  if File.exist? input_yaml_file_name
    File.open(input_yaml_file_name, 'a') { |f|
      f.puts "exit_code: #{exitstatus}"
    }

    if File.exist? output_yaml_file_name
      output_yaml = YAML.load_file(output_yaml_file_name)
      input_yaml = YAML.load_file(input_yaml_file_name)

      # Merge yaml files.
      output_yaml.merge! input_yaml
      File.open(output_yaml_file_name, 'w') { |f|
        f.puts output_yaml.to_yaml
      }
    else
      FileUtils.copy(input_yaml_file_name, output_yaml_file_name)
      `chmod o+w #{output_yaml_file_name}`
    end

    # FileUtils.rm input_yaml_file_name
  else
    puts "Results file: #{s_path}/#{random_string}.yaml does not exist"
  end

end

# Step 5
# Write the results of `docker diff` to OnTrack FS.
# output_path:      Destination path to write files to. This will be a OnTrack FS path.
# diff_result:      String result of `docker diff <container_name>`.
# exec_mode:        Can either be CONSTANTS::BUILD OR CONSTANTS::RUN.
def extract_docker_diff_file(output_path, diff_result, exec_mode)
  File.write("#{output_path}/#{exec_mode}-diff.txt", "docker diff: \n#{!diff_result&.strip&.empty? ? diff_result : 'nothing changed' }")
end

# Step 0, 6
def cleanup_docker_workdir
  return if docker_workdir_path.nil?
  return unless docker_workdir_path.strip.empty? # not nil or empty
  return unless File.exist? docker_workdir_path

  puts "Cleaning HOST_XFS_VOLUME force-recursively: #{docker_workdir_path}/*"
  FileUtils.rm_rf(Dir.glob("#{docker_workdir_path}/*"))
end

def valid_zip_file_param?(params)
  !params['zip_file'].nil? && params['zip_file'].is_a?(Integer) && params['zip_file'] == 1
end

def receive(subscriber_instance, channel, results_publisher, delivery_info, _properties, params)
  puts "*" * 120
  puts "*" * 120
  params = JSON.parse(params)
  return subscriber_instance.client_error!({error: 'PARAM `docker_image_name_tag` is required'}, 400) if params['docker_image_name_tag'].nil?
  return subscriber_instance.client_error!({error: 'PARAM `output_path` is required'}, 400) if params['output_path'].nil?
  return subscriber_instance.client_error!({error: 'PARAM `submission` is required'}, 400) if params['submission'].nil?
  return subscriber_instance.client_error!({error: 'PARAM `assessment` is required'}, 400) if params['assessment'].nil?
  return subscriber_instance.client_error!({error: 'PARAM `timestamp` is required'}, 400) if params['timestamp'].nil?
  return subscriber_instance.client_error!({error: 'PARAM `task_id` is required'}, 400) if params['task_id'].nil?

  if !ENV['RUBY_ENV'].nil? && ENV['RUBY_ENV'] == 'development'
    puts 'Running in development mode.'\
    ' Prepending ROOT_PATH to submission, assessment and output_path params.'
    root_path = ENV['ROOT_PATH']
    params['output_path'] = "#{root_path}#{params['output_path']}"
    params['submission'] = "#{root_path}#{params['submission']}"
    params['assessment'] = "#{root_path}#{params['assessment']}"
  end

  puts "params: #{JSON.pretty_generate(params)}"

  docker_image_name_tag = params['docker_image_name_tag']
  output_path = params['output_path']
  submission = params['submission']
  assessment = params['assessment']
  timestamp = params['timestamp']
  task_id = params['task_id']
  overseer_assessment_id = params['overseer_assessment_id']

  unless task_id.is_a?(Integer)
    subscriber_instance.client_error!({ error: "Invalid task_id: #{task_id}" }, 400)
  end

  unless overseer_assessment_id.is_a?(Integer)
    subscriber_instance.client_error!({ error: "Invalid overseer_assessment_id: #{overseer_assessment_id}" }, 400)
  end

  unless File.exist? submission
    if valid_zip_file_param? params
      subscriber_instance.client_error!({ error: "Zip file not found: #{submission}", overseer_assessment_id: overseer_assessment_id, task_id: task_id, timestamp: timestamp }, 400)
    else
      # By default, Overseer will expect a folder path
      subscriber_instance.client_error!({ error: "Folder not found: #{submission}", overseer_assessment_id: overseer_assessment_id, task_id: task_id, timestamp: timestamp }, 400)
    end
  end

  unless File.exist? assessment
    subscriber_instance.client_error!({ error: "Zip file not found: #{assessment}", overseer_assessment_id: overseer_assessment_id, task_id: task_id, timestamp: timestamp }, 400)
  end

  unless valid_zip? submission
    subscriber_instance.client_error!({ error: "Invalid zip file: #{submission}", overseer_assessment_id: overseer_assessment_id, task_id: task_id, timestamp: timestamp }, 400)
  end

  unless valid_zip? assessment
    subscriber_instance.client_error!({ error: "Invalid zip file: #{assessment}", overseer_assessment_id: overseer_assessment_id, task_id: task_id, timestamp: timestamp }, 400)
  end

  if File.exist? docker_workdir_path
    cleanup_docker_workdir
  end

  if File.exist? docker_execdir_path
    FileUtils.rm_rf docker_execdir_path
  end

  if File.exist? docker_outdir_path
    FileUtils.rm_rf docker_outdir_path
  end

  # TODO: Add correct permissions here
  FileUtils.mkdir_p docker_execdir_path
  FileUtils.mkdir_p docker_outdir_path

  # Clean any output txt and yaml if present
  FileUtils.mkdir_p(output_path) unless File.exist?(output_path)
  FileUtils.rm("#{output_path}/output.txt") if File.exist?("#{output_path}/output.txt")
  FileUtils.rm("#{output_path}/output.yaml") if File.exist?("#{output_path}/output.yaml")

  
  skip_rm = params['skip_rm'] || 0

  # Step 1
  if valid_zip_file_param? params
    # Flatten to ensure submission files are within the root folder not in a task id based subfolder
    extract_zip submission, docker_execdir_path, true, false
  else
    `cp -R #{submission}/. #{docker_execdir_path}`
  end

  # Step 2
  extract_zip assessment, docker_execdir_path, false, true

  random_string = "#{CONSTANTS::BUILD}-#{SecureRandom.hex}"
  exit_code = run_assessment_script_via_docker(
    output_path,
    random_string,
    CONSTANTS::BUILD,
    "chmod u+x /#{CONSTANTS::DOCKER_EXECDIR}/#{CONSTANTS::BUILD}.sh && /#{CONSTANTS::DOCKER_EXECDIR}/#{CONSTANTS::BUILD}.sh /#{CONSTANTS::DOCKER_OUTDIR}/#{random_string}.yaml >> /#{CONSTANTS::DOCKER_OUTDIR}/#{random_string}.txt 2>> /#{CONSTANTS::DOCKER_OUTDIR}/#{random_string}.txt",
    docker_image_name_tag,
    task_id,
    timestamp,
    overseer_assessment_id
  )

  if exit_code == 0
    random_string = "#{CONSTANTS::RUN}-#{SecureRandom.hex}"
    run_assessment_script_via_docker(
      output_path,
      random_string,
      CONSTANTS::RUN,
      "chmod u+x /#{CONSTANTS::DOCKER_EXECDIR}/#{CONSTANTS::RUN}.sh && /#{CONSTANTS::DOCKER_EXECDIR}/#{CONSTANTS::RUN}.sh /#{CONSTANTS::DOCKER_OUTDIR}/#{random_string}.yaml >> /#{CONSTANTS::DOCKER_OUTDIR}/#{random_string}.txt 2>> /#{CONSTANTS::DOCKER_OUTDIR}/#{random_string}.txt",
      docker_image_name_tag,
      task_id,
      timestamp,
      overseer_assessment_id
    )
  end
rescue Subscriber::ClientException => e
  channel.ack(delivery_info.delivery_tag)
  # TODO: Log the error
  puts "Error: #{e.message}"
rescue Subscriber::ServerException => e
  channel.ack(delivery_info.delivery_tag)
  # TODO: Log the error
  puts "Error: #{e.message}"
rescue StandardError => e
  channel.ack(delivery_info.delivery_tag)
  puts "StandardError: #{e.message}"
  subscriber_instance.server_error!({ error: 'Internal server error', overseer_assessment_id: overseer_assessment_id, task_id: task_id, timestamp: timestamp }, 500)
else
  channel.ack(delivery_info.delivery_tag)
  ack_result results_publisher, overseer_assessment_id, task_id, timestamp, output_path
ensure
  if skip_rm != 1
    cleanup_docker_workdir
  else
    puts 'Skipping force removal because skip_rm != 1'
  end
  force_remove_container
end
